### ISSUE TITLE
Impossible to retrieve strings containing certain characters

### ISSUE THREAD

**Issue by sscarduzio** [wg/lettuce#6]
Hi, we store serialized objects inside Redis and - as strings - they typically have special characters inside.

Apparently it's just a parsing error.

We currently use Jedis extensively, but wanted to move to Lettuce for the Async connection feature.
# Test case:
- Install redis 2.6.x
- redis-cli set key "\n\xab\a\n\xf9\x06\n\x0c\b\x01\x12\x06117390\x18\x02"
  OK
- redis-cli get key
  "\n\xab\a\n\xf9\x06\n\x0c\b\x01\x12\x06117390\x18\
- Write this sample Java program (either sync or async)

```
        RedisAsyncConnection<String,String> async = client.connectAsync();
        Future<String> val = async.get("key");
        if (async.awaitAll(val) == true) {
            try {
                System.out.println(val.get());
            } catch (Throwable e) {
                e.printStackTrace();
            }
        }
        client.shutdown();
```
# Observed Output

The value printed is blank. Consistently with different keys containing other serialized objects and using other methods like hmget.

**Comment by wg** [comment]
Hello, lettuce defaults to a UTF-8 key/value codec and I'm guessing those values are not valid UTF-8. It's easy to create your own custom codec, please see https://github.com/wg/lettuce/blob/master/src/test/java/com/lambdaworks/redis/CustomCodecTest.java for an example.

**Comment by sscarduzio** [comment]
Thanks WG, I tried to put together my own binary codec, so that i can query for UTF8 keys and receive values as  as a byte arrays. The encode/decode operations have some println saying what key are they encoding and how many bytes they received back from server. 

See code here: https://gist.github.com/sscarduzio/4996202

And here is how I'm using it:

```
public class LettuceTest {
    public static void main(String[] args) {
        RedisClient client = new RedisClient("staging",7045);
        RedisConnection<String, Object> conn =  client.connect(new BinaryRedisCodec());

        System.out.println("attempting to get a UTF8 compliant key");
        System.out.println(conn.get("key"));

        System.out.println("\nattempting to get a serialized object");
        System.out.println(conn.get("mpme.configuration.v1"));

        conn.shutdown(false);
        client.shutdown();
    }

}
```

And output is:

```
attempting to get a UTF8 compliant key
encoding key
received #bytes 4
[B@79feea8f

attempting to get a serialized object
encoding mpme.configuration.v1
null
```

Notice how in the case of non-UTF value, the BinaryRedisCodec.decode() that contains the "received #bytes: " println is NEVER CALLED. For some external reason, I suppose?

Please advise,

-Simone

**Comment by wg** [comment]
Is it possible that the key you're looking up doesn't actually exist? In that case the decodeValue method would never be called.

**Comment by sscarduzio** [comment]
Hi Wg, thanks for your answer. I also thought like this. But unfortunately the keys are both there.
So I gave these two commands manually:

For the UTF8 value

```
$ redis-cli -p 7045 -h staging get key
"ciao"
```

And the serialized object:

```
$ redis-cli -p 7045 -h staging get mpme.configuration.v1
"\b\x00\x11\x00\x00\x80^9H\xd4A\x1a\x00*U\b\x05\x12(\n\x05.*_US\x12\x1fpartnerId=30&siteID 
```

They are there!

Then I repeated the test this morning and I confirm the test program is returning null.

**HOWEVER** I noticed another thing. Look at what shows from monitor command:

```
$ redis-cli -p 7045 -h staging monitor
OK
1361437136.397650 

1361437149.411381 "GET" "key"
1361437149.444492 "GET" "mpme.configuration.v1\x00\x00"
```

So I replaced the encodeKey method `charset.encode(key).array();` with what I find in UTF8StringCodec.java  `key.getBytes(charset)`  and it worked! :+1: Look:

```
attempting to get a serialized object
encoding 'mpme.configuration.v1'
received #bytes 3409
[B@2085ce5a
```

I'm proud of saying I now can successfully serialize and deserialize my protobuffer objects like a charm :) 

So now I refined the codec class to be support binary input for keys and values.
Have a look: https://gist.github.com/sscarduzio/4996202

Thanks very much for your help!

-Simone
